### PR TITLE
Add most followed channels queries

### PR DESCRIPTION
**Pull request by wyrdrender**

### CODE REVIEW

File: src/aggregates/follows.ts [hunk 1]
```diff
@@ -6,12 +6,24 @@ import {
...
 
+import { ChannelFollowsInfo } from '../entities/ChannelFollowsInfo'
+
 type ChannelEventsAggregationResult = {
   events?: ChannelEvent[]
 }[]
 
 export class FollowsAggregate implements GenericAggregate<ChannelEvent> {
   private channelFollowsMap: Record<string, number> = {}
+  private allChannelFollowsEvents: Partial<UnsequencedChannelEvent>[] = []
```
Comment (kdembler) [inline src/aggregates/follows.ts:17]: Let's not pluralize `follows` here: `allChannelFollowEvents`

File: src/aggregates/follows.ts [hunk 2]
```diff
@@ -6,12 +6,24 @@ import {
...
+  private allChannelFollows: ChannelFollowsInfo[] = []
+
+  private addOrUpdateFollows(array: ChannelFollowsInfo[], id: string, event: ChannelEventType): void {
```
Comment (kdembler) [inline src/aggregates/follows.ts:20]: Let's use `eventType` for name

File: src/aggregates/follows.ts [hunk 3]
```diff
@@ -6,12 +6,24 @@ import {
...
+    const i = array.findIndex((element) => element.id === id)
+    if (i > -1) {
+      if (!array[i].follows && event === ChannelEventType.UnfollowChannel) return
+      array[i].follows = event === ChannelEventType.FollowChannel ? array[i].follows + 1 : array[i].follows - 1
```
Comment (kdembler) [inline src/aggregates/follows.ts:24]: You need to make sure to not get to negative values

File: src/aggregates/follows.ts [hunk 4]
```diff
@@ -6,12 +6,24 @@ import {
...
+    } else array.push({ id, follows: 1 })
```
Comment (kdembler) [inline src/aggregates/follows.ts:25]: This doesn't handle unfollow event properly

File: src/resolvers/followsInfo.ts [hunk 5]
```diff
@@ -31,6 +54,22 @@ export class ChannelFollowsInfosResolver {
     return getFollowsInfo(channelId, ctx)
   }
 
+  @Query(() => [ChannelFollowsInfo], { description: 'Get most followed list of channels' })
```
Comment (kdembler) [inline src/resolvers/followsInfo.ts:57]:
```suggestion
  @Query(() => [ChannelFollowsInfo], { description: 'Get list of most followed channels' })
```

File: src/resolvers/followsInfo.ts [hunk 6]
```diff
@@ -31,6 +54,22 @@ export class ChannelFollowsInfosResolver {
...
+  async mostFollowedChannels(
+    @Args() { period, limit }: MostFollowedChannelsArgs,
+    @Ctx() ctx: OrionContext
+  ): Promise<ChannelFollowsInfo[]> {
+    return mapMostFollowedArray(buildMostFollowedChannelsArray(ctx, period), limit)
+  }
+
+  @Query(() => [ChannelFollowsInfo], { nullable: true, description: 'Get most followed list of channels of all time' })
```
Comment (kdembler) [inline src/resolvers/followsInfo.ts:65]:
```suggestion
  @Query(() => [ChannelFollowsInfo], { nullable: true, description: 'Get list of most followed channels of all time' })
```

**Review by kdembler — CHANGES_REQUESTED** [review]

File: src/aggregates/follows.ts (hunk 3 above)
Comment (wyrdrender) [reply src/aggregates/follows.ts:24]: But there is a check for that in line 23

Comment (kdembler) [reply src/aggregates/follows.ts:24]: Okay, missed that

File: src/aggregates/follows.ts [hunk 7]
```diff
@@ -6,12 +6,39 @@ import {
...
+
+  private addOrRemoveFollowEvent(eventType: ChannelEventType, { channelId, timestamp }: UnsequencedChannelEvent): void {
+    if (eventType === ChannelEventType.FollowChannel) {
+      this.allChannelFollowEvents = [...this.allChannelFollowEvents, { channelId, timestamp }]
+    }
+    if (eventType === ChannelEventType.UnfollowChannel) {
+      const followEvent = this.allChannelFollowEvents.find((item) => item.channelId === channelId)
+      if (followEvent) {
+        this.allChannelFollowEvents.splice(this.allChannelFollowEvents.indexOf(followEvent), 1)
+      }
```
Comment (kdembler) [inline src/aggregates/follows.ts:39]: I think we could use `findIndex` instead of `find` followed by `indexOf`?

File: src/aggregates/follows.ts [hunk 8]
```diff
@@ -33,10 +33,10 @@ export class FollowsAggregate implements GenericAggregate<ChannelEvent> {
...
     }
     if (eventType === ChannelEventType.UnfollowChannel) {
-      const followEvent = this.allChannelFollowEvents.find((item) => item.channelId === channelId)
-      if (followEvent) {
-        this.allChannelFollowEvents.splice(this.allChannelFollowEvents.indexOf(followEvent), 1)
-      }
+      this.allChannelFollowEvents.splice(
+        this.allChannelFollowEvents.findIndex((item) => item.channelId === channelId),
+        1
+      )
```
Comment (kdembler) [inline src/aggregates/follows.ts:39]: We still need to protect against missing event. Currently if the event won't be find, you will remove the last element of the array

**Review by kdembler — APPROVED** [review]

**Pull request merged**